Add sugar for handling borrowed and owned types

Manishearth opened the pull request:
Implements the changes outlined in https://github.com/servo/servo/pull/12826#discussion_r75074985

<s>Also gets things ready for the Unique/Borrowed bindings</s>

WIP for borrowed and unique in the same PR. Still need to convert all the rest of the gecko types to use the new wrappers.

r? @emilio

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12991)

<!-- Reviewable:end -->

Manishearth committed:
Make Arc helpers into methods

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/gecko.mako.rs, components/style/gecko_conversions.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

Manishearth committed:
Allow for regular non-arc borrowed types, use for StyleSet

Manishearth commented:
Do we ever have non-arc borrowed types which can be null? Here I'm treating them as `&T`, but perhaps we should be using `Borrowed<T>` instead

Manishearth commented:
The same question stands for owned values. We don't seem to have any null owned values right now, but we may get them later?

Manishearth committed:
Add bindings for owned types, use for servostyleset

Manishearth commented:
Gecko side at https://github.com/Manishearth/gecko-dev/tree/more-arc-safety

I'll upload to bugzilla once I get confirmation that this approach is right.

Manishearth committed:
Rename sugar::refptr to sugar::ownership

bholley commented:
Per today's meeting, I think it would be nice to support optional nullability in the type system if that's not too much work (but have the default assume non-null).

Manishearth commented:
Yep. I plan to do that via OwnedNullable and BorrowedNullable. Don't yet see any nullable types in the FFI calls, though.

Another thing I might add is the ability to return `Strong<T>` instead of already_addrefed in `Gecko_*`. While we've been using `Strong<T>` exclusively as a stand-in for `Arc<T>`, the finer-grained traits make it possible to use `Strong<T>` for both arc and non-arc strong refcounted types.

Manishearth committed:
Add MaybeOwned, and use for ServoNodeData. More docs.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #12986) made this pull request unmergeable. Please resolve the merge conflicts.

Manishearth committed:
Handle RawGeckoElement and friends, add support for types which are used in both maybe-null and non-null forms

Manishearth commented:
Ugh, this got stickier. We deal with RawGeckoNode and friends in both non-null and maybe-null ways. Looks like I will need to have MaybeBorrowed typedefs on the gecko side too.

Manishearth commented:
Updated to handle RawGeckoNode/Element/Document

Manishearth commented:
cc @mystor

mystor commented:
Review status: 0 of 9 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 15 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1BH0IbThZJzudBm3k:-KQ1BH0JZKtQNLSwxH4I:b-mkp4rx) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L15)):*
> ```Rust
> /// use HasArcFFI or similar traits to define it
> pub unsafe trait HasFFI : Sized {
>     type FFIType: Sized;
> ```

It seems to me like it might be easier to have `Borrowed<>` and `Strong<>` be generic over a type which implements `HasArcFFI`, like so:

pub struct Borrowed<'a, T: HasArcFFI + 'a> {
    ptr: *const T::FFIType,
    _marker: PhantomData<&'a T::FFIType>,
}

The methods which currently require that awkward <U> type parameter could then avoid using that type parameter at all. I think that that would simplify the type signatures of the methods on `Borrowed` and `Strong` quite a bit.

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 27 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ15QmdgapfsGWb6G7r:-KQ15QmeZeoz8qZkP6G5:bb62zfk) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L27)):*
> ```Rust
>     /// Artificially increments the refcount of a borrowed Arc over FFI.
>     unsafe fn addref(ptr: Borrowed<Self::FFIType>) {
>         forget(ptr.as_arc::<Self>().clone())
> ```

Again, this method assumes that ptr is non-null, while release() assumes that ptr is null. Why not make them match?

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 57 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ19wZ5yQNmBpMfOdyX:-KQ19wZ5yQNmBpMfOdyY:b-mbq65e) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L57)):*
> ```Rust
> 
> impl<'a, T> Borrowed<'a, T> {
>     pub fn is_null(&self) -> bool {
> ```

It really doesn't matter, but because Borrowed is `Copy`, this can probably just be done by value to avoid a pointer indirection.

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 61 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1604I010iTwQ15y0X:-KQ1604I010iTwQ15y0Y:b78aqr4) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L61)):*
> ```Rust
>     }
> 
>     pub fn as_arc_opt<U>(&self) -> Option<&Arc<U>> where U: HasArcFFI<FFIType = T> {
> ```

This extra type parameter seems inconvenient, especially because it seems like we should be able to get from the FFIType back to the base type easily. Are there FFI types which have multiple rust types?

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 73 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ16RaRMVGbPHHDRxUB:-KQ16RaRMVGbPHHDRxUC:b7a9z8s) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L73)):*
> ```Rust
>     /// Converts a borrowed FFI reference to a borrowed Arc.
>     /// Panics on null
>     pub fn as_arc<U>(&self) -> &Arc<U> where U: HasArcFFI<FFIType = T> {
> ```

Why do we have this method and the above method, instead of just having `as_arc(&self) -> Option<&Arc<U>>`, and `as_arc().unwrap()`?

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 75 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ17ywAkMmIaqWTxM5B:-KQ17ywEGBONVt-JUwQm:bkx8avt) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L75)):*
> ```Rust
>     pub fn as_arc<U>(&self) -> &Arc<U> where U: HasArcFFI<FFIType = T> {
>         self.as_arc_opt().unwrap()
>     }
> ```

You can also have null `Borrowed<T>`s. Add a `null() -> Borrowed<T>` method.

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 80 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ192RSvCAeMAOCX6aF:-KQ192RSvCAeMAOCX6aG:betg5m4) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L80)):*
> ```Rust
> #[repr(C)]
> /// Gecko-FFI-safe Arc (T is an ArcInner).
> /// This can be null.
> ```

Add a note that this type, if dropped, will cause memory to leak, as it has no `Drop` impl (this should be fixable when we get rid of drop flags).

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 81 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1DwHVk5GCQXKMQJDu:-KQ17m2REJonjiUsISuT:bunjy2o) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L81)):*
> ```Rust
> /// Gecko-FFI-safe Arc (T is an ArcInner).
> /// This can be null.
> pub struct Strong<T> {
> ```

Usually I would refer to them as `Strong` and `Weak` pointers, or `Owned` and `Borrowed` pointers. It feels like the names in this module are mixing metaphors a bit.

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 84 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ188tPV4Uyi8nz0Zc9:-KQ188tTndz0tFGM9iN6:borymso) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L84)):*
> ```Rust
>     ptr: *const T,
>     _marker: PhantomData<T>,
> }
> ```

`Borrowed<T>` is Copy and Clone. This method could be `Clone` and `addref` itself when cloned. Not sure if that would be super useful, as this type will probably be `into_arc`-ed almost immediately after entering rust code.

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 93 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ17OZkumRqq71WYC9Y:-KQ17OZkumRqq71WYC9Z:b-vnzt8b) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L93)):*
> ```Rust
>     /// Given a non-null strong FFI reference, converts it into an Arc.
>     /// Panics on null.
>     pub fn into_arc<U>(self) -> Arc<U> where U: HasArcFFI<FFIType = T> {
> ```

You at least need the option for a `into_arc_opt()`, but again I think I would prefer `into_arc(self) -> Option<Arc<U>>` and then people who know that `self` is non-null can just `.unwrap()` the result. 

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 99 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ17GYWBwEIbcuRLiWU:-KQ17GYWBwEIbcuRLiWV:bubvh8m) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L99)):*
> ```Rust
> 
>     /// Produces a null strong FFI reference
>     pub fn null_strong() -> Self {
> ```

This method is on the Strong type directly, you can probably just call it null()

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 102 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ18qED9Hmm0GWvLqd5:-KQ18qED9Hmm0GWvLqd6:bgk6epa) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L102)):*
> ```Rust
>         unsafe { transmute(ptr::null::<T>()) }
>     }
> }
> ```

There should be some way to get a Borrowed<'a, T> from a &'a Strong<T>. This would allow for, for example, explicit `addref` or `release` calls with `Strong<T>`.

---

*[ports/geckolib/gecko_bindings/sugar/refptr.rs, line 115 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ175qB2O-4II4JBLJA:-KQ175qCSA6ml2Epql8_:btsg3um) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/refptr.rs#L115)):*
> ```Rust
>     type Inner = T;
>     fn into_strong(self) -> Strong<T::FFIType> {
>         unsafe {transmute(self)}
> ```

Spacing

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12991)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 0 of 9 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 15 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1BH0IbThZJzudBm3k:-KQ1HrxXXdTsjZBbC2JX:b-n8lx4a) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L15)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> It seems to me like it might be easier to have `Borrowed<>` and `Strong<>` be generic over a type which implements `HasArcFFI`, like so:
> 
> pub struct Borrowed<'a, T: HasArcFFI + 'a> {
>     ptr: *const T::FFIType,
>     _marker: PhantomData<&'a T::FFIType>,
> }
> 
> The methods which currently require that awkward <U> type parameter could then avoid using that type parameter at all. I think that that would simplify the type signatures of the methods on `Borrowed` and `Strong` quite a bit.
</details>
Borrowed is used for both arc-y FFI and non-arc-y FFI (and the behavior of the two is different).

Also, where clauses don't work that way without introducing a second phantom type parameter. T is not HasArcFFI, T is HasArcFFI::FFIType.

Of course I could invert HasArcFFI and implement it on the FFI types, but coherence is an issue. The FFI types are always defined in gecko_bindings (this crate), but the Servo types can be defined in style or geckolib, both which depend on this. The impl isn't coherent.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 27 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ15QmdgapfsGWb6G7r:-KQ1IWoZd_9fzi24FAlQ:bgbbn8a) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L27)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> Again, this method assumes that ptr is non-null, while release() assumes that ptr is null. Why not make them match?
</details>
Because we weren't doing that in the past. I think unconditional release() is called in the destructors, but addref is only called when we know it's non-null.

But I changed it to forget as_arc_opt().clone() instead.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 57 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ19wZ5yQNmBpMfOdyX:-KQ1JIbw0dq1-pzcbjpR:bw4y1w9) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L57)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> It really doesn't matter, but because Borrowed is `Copy`, this can probably just be done by value to avoid a pointer indirection.
</details>
Yeah. All the non-arc borrowed methods should take `self`. Changed `as_ffi` too, because that was wrong.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 61 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1604I010iTwQ15y0X:-KQ1JK93UYKtR-5ohHEl:bg5eo8h) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L61)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> This extra type parameter seems inconvenient, especially because it seems like we should be able to get from the FFIType back to the base type easily. Are there FFI types which have multiple rust types?
</details>
No, but there's no way of teaching that to the type system. A second tra

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 73 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ16RaRMVGbPHHDRxUB:-KQ1Jb6LZCcnJP4sdgZk:b-oh2dxf) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L73)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> Why do we have this method and the above method, instead of just having `as_arc(&self) -> Option<&Arc<U>>`, and `as_arc().unwrap()`?
</details>
Convenience, `as_arc` is used often. Also lets you do things like `.map(Borrowed::as_arc)`.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 75 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ17ywAkMmIaqWTxM5B:-KQ1JjMbqgbudZ3RpEPe:b-6648q) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L75)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> You can also have null `Borrowed<T>`s. Add a `null() -> Borrowed<T>` method.
</details>
Dangling lifetimes!

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 81 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1DwHVk5GCQXKMQJDu:-KQ1K-1jQI_M8tMLNf3W:b37g4sp) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L81)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> Usually I would refer to them as `Strong` and `Weak` pointers, or `Owned` and `Borrowed` pointers. It feels like the names in this module are mixing metaphors a bit.
</details>
This is because Borrowed is used for both nullable &T and nullable &ArcInner

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 102 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ18qED9Hmm0GWvLqd5:-KQ1KznIZgkBeo_kMyVL:bktx42e) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L102)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> There should be some way to get a Borrowed<'a, T> from a &'a Strong<T>. This would allow for, for example, explicit `addref` or `release` calls with `Strong<T>`.
</details>
Servo never wants to call addref or release. Strong isn't meant to be held on to, you either convert it to an arc and addref there, or pass back to gecko.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12991)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 0 of 11 files reviewed at latest revision, 13 unresolved discussions.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 84 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ188tPV4Uyi8nz0Zc9:-KQ1MUUw-mvvVO4Gjgjb:b-l0qm7i) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L84)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> `Borrowed<T>` is Copy and Clone. This method could be `Clone` and `addref` itself when cloned. Not sure if that would be super useful, as this type will probably be `into_arc`-ed almost immediately after entering rust code.
</details>
not sure which method is being talked about, reviewable lost the location

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12991)*
<!-- Sent from Reviewable.io -->

Manishearth committed:
Review fixes and bindings resync

bors-servo commented:
:umbrella: The latest upstream changes (presumably #12911) made this pull request unmergeable. Please resolve the merge conflicts.

mystor commented:
Review status: 0 of 11 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[ports/geckolib/glue.rs, line 127 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQTgUsFx8NigDLuk19s:-KQTgUsFx8NigDLuk19t:b1pv26q) ([raw file](https://github.com/servo/servo/blob/685448a64ff1f368f6ed34bd69e07a082e4b44ca/ports/geckolib/glue.rs#L127)):*
> ```Rust
> #[no_mangle]
> pub extern "C" fn Servo_RestyleDocument(doc: RawGeckoDocumentBorrowed, raw_data: RawServoStyleSetBorrowedMut) -> () {
>     let document = unsafe { GeckoDocument(doc) };
> ```

Why does GeckoDocument require an unsafe block to construct?

---

*[ports/geckolib/glue.rs, line 272 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQTh6AKjvxgmtX98Vmx:-KQTh6AKjvxgmtX98Vmy:bbm89iu) ([raw file](https://github.com/servo/servo/blob/685448a64ff1f368f6ed34bd69e07a082e4b44ca/ports/geckolib/glue.rs#L272)):*
> ```Rust
> #[no_mangle]
> pub extern "C" fn Servo_ComputedValues_GetForPseudoElement(parent_style: ServoComputedValuesMaybeBorrowed,
>                                                           match_element: RawGeckoElementBorrowed,
> ```

Indentation seems a bit odd here

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 15 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1BH0IbThZJzudBm3k:-KQ1qsr9gdXEIULHfzCw:bachlf5) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L15)):*
<details><summary><i>Previously, Manishearth (Manish Goregaokar) wrote…</i></summary>
> Borrowed is used for both arc-y FFI and non-arc-y FFI (and the behavior of the two is different).
> 
> Also, where clauses don't work that way without introducing a second phantom type parameter. T is not HasArcFFI, T is HasArcFFI::FFIType.
> 
> Of course I could invert HasArcFFI and implement it on the FFI types, but coherence is an issue. The FFI types are always defined in gecko_bindings (this crate), but the Servo types can be defined in style or geckolib, both which depend on this. The impl isn't coherent.
</details>
OK. We talked some more and we can't do what I was suggesting until something like #13038 is merged.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 75 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ17ywAkMmIaqWTxM5B:-KQ1rZurjI-E7FjeejJl:bvymbem) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L75)):*
<details><summary><i>Previously, Manishearth (Manish Goregaokar) wrote…</i></summary>
> Dangling lifetimes!
</details>
We can have `fn null() -> Borrowed<'static, T>`. It is technically correct, the data behind `null` will last forever (because it doesn't exist). I don't think that this is a problem.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 81 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1DwHVk5GCQXKMQJDu:-KQ1sll0laQyoj2fF1DO:b6qw0bg) ([raw file](https://github.com/servo/servo/blob/7437f8894693b1a86ac097c2537ab2008758ae3f/ports/geckolib/gecko_bindings/sugar/ownership.rs#L81)):*
<details><summary><i>Previously, Manishearth (Manish Goregaokar) wrote…</i></summary>
> This is because Borrowed is used for both nullable &T and nullable &ArcInner
</details>
OK, So Borrowed is also used for HasSimpleFFI? That makes sense.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 153 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1rrUQWFGmM5BasgFV:-KQ1rrURyTUa1ZsKJPHI:batpqdi) ([raw file](https://github.com/servo/servo/blob/685448a64ff1f368f6ed34bd69e07a082e4b44ca/ports/geckolib/gecko_bindings/sugar/ownership.rs#L153)):*
> ```Rust
>     #[inline]
>     /// Borrowed<ServoType> -> Borrowed<GeckoType>
>     pub fn as_ffi(self) -> Borrowed<'a, <Self as HasFFI>::FFIType> where Self: HasSimpleFFI {
> ```

Shouldn't it be `Borrowed<'a, <T as HasFFI>::FFIType> where T: HasSimpleFFI`?

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 201 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQTiDZIwQ6e1_EgFssG:-KQTiDZJxQ88ZB3yqLMd:b-468axt) ([raw file](https://github.com/servo/servo/blob/685448a64ff1f368f6ed34bd69e07a082e4b44ca/ports/geckolib/gecko_bindings/sugar/ownership.rs#L201)):*
> ```Rust
> }
> 
> // technically not
> ```

technically not what? Maybe just write "Deref to Borrowed so that BorrowedMut can be used where Borrowed is"

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 260 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQTibFmrjDhhtwaJ_hJ:-KQTibFmrjDhhtwaJ_hK:bskxl56) ([raw file](https://github.com/servo/servo/blob/685448a64ff1f368f6ed34bd69e07a082e4b44ca/ports/geckolib/gecko_bindings/sugar/ownership.rs#L260)):*
> ```Rust
>     ///
>     /// Arc<ServoType> -> Strong<GeckoType>
>     fn into_strong(self) -> Strong<<Self::Inner as HasFFI>::FFIType>;
> ```

I'm not sure I like this name, as it is converting to an ffi type, but doesn't really make that clear. Maybe `into_strong_ffi` and `as_borrowed_ffi`? They're longer, which sucks, but I feel like they make the intention more clear

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 315 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1yhIlFOrIoDUUxsb0:-KQ1yhImWkPd235Du7Ey:b-5c505n) ([raw file](https://github.com/servo/servo/blob/685448a64ff1f368f6ed34bd69e07a082e4b44ca/ports/geckolib/gecko_bindings/sugar/ownership.rs#L315)):*
> ```Rust
> /// Gecko-FFI-safe owned pointer
> /// Can be null
> pub struct MaybeOwned<T> {
> ```

This one also doesn't free its backing memory on drop.

You have this `Owned`/`MaybeOwned` pairing. Is there any particular reason why you don't also have `Borrowed`/`MaybeBorrowed` and `Strong`/`MaybeStrong` etc? 

Also, in Rust we call the type `Option`, so maybe `OptionalOwned` would be better?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12991)*
<!-- Sent from Reviewable.io -->

mystor commented:
Reviewable seems to have dumped a ton of random stuff into that comment above - I didn't really have much to say, just a couple of silly minor nits. 

r+ from me

emilio commented on ports/geckolib/gecko_bindings/sugar/ownership.rs:
```diff
@@ -0,0 +1,340 @@
+/// Gecko-FFI-safe mutably borrowed type
+/// This can be null.
+pub struct BorrowedMut<'a, T: 'a> {
+    ptr: *mut T,
```
Seems to me that this could be `Option<*mut T>`? Then most types could use `as_ref()`, or similar.

Manishearth replied:
`Option` isn't `repr(C)`, and I'm wary of relying on that optimization. Coherence makes this worse. I'll revisit it once we've merged the crates.

emilio commented on ports/geckolib/gecko_bindings/sugar/ownership.rs:
```diff
@@ -0,0 +1,340 @@
+/// This can be null.
+/// Leaks on drop. Please don't drop this.
+pub struct Strong<T> {
+    ptr: *const T,
```
Ditto?

emilio commented on ports/geckolib/gecko_bindings/sugar/ownership.rs:
```diff
@@ -0,0 +1,340 @@
+#[repr(C)]
+/// Gecko-FFI-safe Arc (T is an ArcInner).
+/// This can be null.
+/// Leaks on drop. Please don't drop this.
```
Also, we don't have a rust compiler with MIR enabled yet, so maybe this can't be done right now, but maybe leaving a todo comment adding a `Drop` impl that asserts could be nice.

emilio commented on ports/geckolib/wrapper.rs:
```diff
@@ -342,47 +339,32 @@ impl<'ln> TNode for GeckoNode<'ln> {
-    element: *mut RawGeckoElement,
-    chain: PhantomData<&'le ()>,
-}
+pub struct GeckoElement<'le>(pub &'le RawGeckoElement);
```
I'd sort of prefer the `.element` field name to make it similar to servo's implementation (here and in `Node`/`Document`. But I guess I'm fine with this too.

emilio commented on ports/geckolib/binding_tools/regen.py:
```diff
@@ -307,16 +334,60 @@ def build(objdir, target_name, debug, debugger, kind_name=None,
+            flags.append("-raw-line")
+            flags.append("pub type {0}Borrowed<'a> = &'a {0};".format(ty))
+            flags.append("-blacklist-type")
+            flags.append("{}MaybeBorrowed".format(ty))
```
Also, this needs a rebase for the bindgen change.

emilio commented:
I've been sort of out of the loop with all this type safety effort, but I'm ok with this changes. I think I'm missing a lot of background though, so happy to redirect the review to @mystor or @bholley.

I only pointed out a few things that should be possible (like using directly `Option<*mut T>` instead of raw `*mut T` then checking for null, but otherwise I don't have a lot more to say :)

Manishearth commented:
Review status: 0 of 11 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 153 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1rrUQWFGmM5BasgFV:-KQU63gMa_elcqAlapg9:bp9znli) ([raw file](https://github.com/servo/servo/blob/685448a64ff1f368f6ed34bd69e07a082e4b44ca/ports/geckolib/gecko_bindings/sugar/ownership.rs#L153)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> Shouldn't it be `Borrowed<'a, <T as HasFFI>::FFIType> where T: HasSimpleFFI`?
</details>
I don't know which part of the code you're talking about here

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 315 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQ1yhIlFOrIoDUUxsb0:-KQU68tVirD85ZVeYUj8:bw77g7l) ([raw file](https://github.com/servo/servo/blob/685448a64ff1f368f6ed34bd69e07a082e4b44ca/ports/geckolib/gecko_bindings/sugar/ownership.rs#L315)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> This one also doesn't free its backing memory on drop.
> 
> You have this `Owned`/`MaybeOwned` pairing. Is there any particular reason why you don't also have `Borrowed`/`MaybeBorrowed` and `Strong`/`MaybeStrong` etc? 
> 
> Also, in Rust we call the type `Option`, so maybe `OptionalOwned` would be better?
</details>
`Borrowed` is just `&T`. I might rename them later, but I'd rather do it in a separate PR, as this keeps bitrotting and the separate PR can be made on purely the servo side.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12991)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 0 of 11 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 201 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQTiDZIwQ6e1_EgFssG:-KQUX96SB9ALIDvOLx3k:b-fnaxdn) ([raw file](https://github.com/servo/servo/blob/685448a64ff1f368f6ed34bd69e07a082e4b44ca/ports/geckolib/gecko_bindings/sugar/ownership.rs#L201)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> technically not what? Maybe just write "Deref to Borrowed so that BorrowedMut can be used where Borrowed is"
</details>
"technically not how we're supposed to use Deref".  But it helps us avoid duplication and it's just a style nit, so meh

---

*[ports/geckolib/gecko_bindings/sugar/ownership.rs, line 260 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12991#-KQTibFmrjDhhtwaJ_hJ:-KQUXIh6SHdojOV-GmS0:bfsgjzd) ([raw file](https://github.com/servo/servo/blob/685448a64ff1f368f6ed34bd69e07a082e4b44ca/ports/geckolib/gecko_bindings/sugar/ownership.rs#L260)):*
<details><summary><i>Previously, mystor (Michael Layzell) wrote…</i></summary>
> I'm not sure I like this name, as it is converting to an ffi type, but doesn't really make that clear. Maybe `into_strong_ffi` and `as_borrowed_ffi`? They're longer, which sucks, but I feel like they make the intention more clear
</details>
as_ always means borrowed, into_ always means owned, in most Rust APIs

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12991)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
Updated

Manishearth commented:
Gecko PR at https://bugzilla.mozilla.org/show_bug.cgi?id=1299392

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13134) made this pull request unmergeable. Please resolve the merge conflicts.

Manishearth commented:
Did some renaming after review comments on the gecko side. r? @emilio

emilio commented on ports/geckolib/binding_tools/regen.py:
```diff
@@ -315,16 +341,61 @@ def build(objdir, target_name, debug, debugger, kind_name=None,
+            flags.append("--blacklist-type")
+            flags.append("{}BorrowedOrNull".format(ty))
+            flags.append("--raw-line")
+            flags.append("pub type {0}BorrowedOrNull<'a> = ::sugar::ownership::Borrowed<'a, {0}>;".format(ty))
```
I think we're hugely abusing --raw-line, which was supposed to be a temporary workaround, but... meh.

Manishearth replied:
We could use `include!` instead, and have regen.py generate that file.

I plan to make bingen read from a json file instead of arguments at some point.

emilio commented:
Can you squash? r=me,mystor with that

emilio commented on ports/geckolib/wrapper.rs:
```diff
@@ -345,14 +345,14 @@ impl<'ln> TNode for GeckoNode<'ln> {
         if let GeckoChildrenIterator::GeckoIterator(ref it) = *self {
             unsafe {
-                Gecko_DropStyleChildrenIterator(*it as *const _ as *mut _);
+                Gecko_DropStyleChildrenIterator(ptr::read(it as *const _));
```
Is this `ptr::read` really needed?

Manishearth replied:
`drop` borrows, and we can't move out of a borrow.

bholley commented:
This should be fine to land now that bug 1299392 has hit central. Once it does I'll update incubator.

Manishearth commented:
@bors-servo r=mystor,emilio

bors-servo commented:
:pushpin: Commit f72cd7f has been approved by `mystor,emilio`

<!-- @bors-servo r=mystor,emilio f72cd7ffbc4997fc49394abbc3772652a8623f61 -->

bors-servo commented:
:hourglass: Testing commit f72cd7f with merge 509a293...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2948)

highfive commented:
▶ TIMEOUT [expected PASS] /_mozilla/css/iframe/hide_layers2.html

Manishearth commented:
@bors-servo retry
 - #13059

bors-servo commented:
:hourglass: Testing commit f72cd7f with merge 927cd8e...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2775), [arm64](http://build.servo.org/builders/arm64/builds/2792), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2845), [linux-rel](http://build.servo.org/builders/linux-rel/builds/3010), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2855), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2846), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2950), [windows-dev](http://build.servo.org/builders/windows-dev/builds/741)
